makeservices/makexmpp-real

```diff
@@ -0,0 +1,250 @@
+compromising the shell account password. Note that the XMPP server should also
+prevent the user from later changing the password.
+
+The admin credentials are read from a config file; the path to it is stored in
+the CONF_FILE global. This file must be readable only by the setuid user! The
+Python configparser module is used to parse the config file, so the format is
+basically the Windows INI format.
+
+Most of the code in this file was adapted from a SleekXMPP example program, see
+https://github.com/fritzy/SleekXMPP/blob/master/examples/admin_commands.py
+"""
+import os
+import random
+import string
+import sys
+from configparser import ConfigParser
+from textwrap import dedent
+
+import sleekxmpp
+from ocflib.misc.mail import send_problem_report
+
+
+CONF_FILE = '/opt/share/makexmpp/makexmpp.conf'
+JID_DOMAIN = 'ocf.berkeley.edu'
+
+PW_LENGTH = 24
+
+
+def read_config():
+    """Fetches the server admin JID and password from the config in
+    '/opt/share/makeservices'."""
+    conf = ConfigParser()
+    conf.readfp(open(CONF_FILE))
+    admin_jid = conf.get('makexmpp', 'jid')
+    admin_pw = conf.get('makexmpp', 'passwd')
+    return admin_jid, admin_pw
+
+
+def generate_password(length):
+    r = random.SystemRandom()
+    return ''.join(
+        r.choice(string.ascii_letters + string.digits)
+        for _ in range(length)
+    )
+
+
+def intro_prompt():
+    print(dedent(
+        """
+        This program will create an XMPP account, if one does not already exist.
+        A randomly-generated password will be generated for this account and
+        displayed on-screen. Please make sure you are in an environment where
+        nobody else will see it when it appears.
+
+        You can always re-run this command to reset the password if you lose it.
+
+        If you are ready to continue, type 'yes'.
+        Typing anything other than yes will abort this script.
+        """
+    ))
+    return input('Continue? ') == 'yes'
+
+
+class XMPPUserPassword(sleekxmpp.ClientXMPP):
+    def __init__(self, jid, password, newuser, newpassword):
+        sleekxmpp.ClientXMPP.__init__(self, jid, password)
+
+        self._newuser = newuser
+        self._newjid = newuser + '@' + JID_DOMAIN
+        self._newpassword = newpassword
+
+        self.add_event_handler('session_start', self.start)
+
+    def start(self, event):
+        """
+        Process the session_start event. We first try to change the user's
+        password. If the account does not exist, we then create it.
+
+        `event` is an empty dictionary. The session_start event does not
+        provide any additional data.
+        """
+
+        def command_error(iq, session):
+            self['xep_0050'].terminate_command(session)
+            self.disconnect()
+
+            condition = iq['error']['condition']
+            errtext = iq['error']['text']
+            raise Exception('{}: {}'.format(condition, errtext))
```

Comment:
I don't think this is a huge deal, since this is only seen by the emailed problem report in the `root@` mailing list. We'll see the line number which is more important anyways.